### PR TITLE
Form-specific access control for standard encounter forms.

### PR DESCRIPTION
**Pull request by sunsetsystems**
This is similar to and builds on my PR #473 on 2017-02-08, but deals with standard encounter forms rather than layout-based. In the forms admin page you can specify an ACO for each form, and appropriate checking is done where menus and buttons are presented.

Reviews/comments appreciated.

### CODE REVIEW

File: interface/forms_admin/forms_admin.php [hunk 1]
```diff
@@ -37,103 +42,107 @@
...
+    else if (preg_match('/aco_spec_(\d+)/', $key, $matches)) {
+      sqlQuery("update registry set aco_spec = ? where id = ?", array($val, $matches[1]));
+    }
+  }
 ?>
 
-
 <?php //ERROR REPORTING
 if ($err)
 	echo "<span class=bold>$err</span><br><br>\n";
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: text() around $err

File: interface/forms_admin/forms_admin.php [hunk 2]
```diff
@@ -37,103 +42,107 @@
...
-	if ($color=="#CCCCCC")
-		$color="#999999";
-	else
-		$color="#CCCCCC";
+  $priority_category = sqlQuery("select priority, category, nickname, aco_spec from registry where id = ?",
+    array($registry['id']));
+?>
+  <tr>
+    <td bgcolor="<?php echo $color; ?>" width="2%">
+      <span class='text'><?php echo $registry['id']; ?></span>
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: text()

File: interface/forms_admin/forms_admin.php [hunk 3]
```diff
@@ -37,103 +42,107 @@
...
+    </td>
+    <td bgcolor="<?php echo $color; ?>" width="30%">
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: attr()
(I realize this is a bit on the anal side, but never know if $color will become a option from user selection in the future :) )

File: interface/forms_admin/forms_admin.php [hunk 4]
```diff
@@ -37,103 +42,107 @@
...
+      <span class='bold'><?php echo xl_form_title($registry['name']); ?></span>
+    </td>
+    <?php
+      if ($registry['sql_run'] == 0)
+        echo "<td bgcolor='$color' width='10%'><span class='text'>".xl('registered')."</span>";
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: xlt()

File: interface/forms_admin/forms_admin.php [hunk 5]
```diff
@@ -37,103 +42,107 @@
...
+      elseif ($registry['state'] == "0")
+        echo "<td bgcolor='#FFCCCC' width='10%'><a class='link_submit' href='./forms_admin.php?id={$registry['id']}&method=enable'>".xl('disabled')."</a>";
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: xlt()

File: interface/forms_admin/forms_admin.php [hunk 6]
```diff
@@ -37,103 +42,107 @@
...
+      else
+        echo "<td bgcolor='#CCFFCC' width='10%'><a class='link_submit' href='./forms_admin.php?id={$registry['id']}&method=disable'>".xl('enabled')."</a>";
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: xlt()

File: interface/forms_admin/forms_admin.php [hunk 7]
```diff
@@ -37,103 +42,107 @@
...
+    ?></td>
+    <td bgcolor="<?php echo $color; ?>" width="10%">
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: attr()

File: interface/forms_admin/forms_admin.php [hunk 8]
```diff
@@ -37,103 +42,107 @@
...
+      <span class='text'><?php
+      if ($registry['unpackaged'])
+        echo xlt('PHP extracted');
+      else
+        echo xlt('PHP compressed');
+      ?></span>
+    </td>
+    <td bgcolor="<?php echo $color; ?>" width="10%">
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: attr()

File: interface/forms_admin/forms_admin.php [hunk 9]
```diff
@@ -37,103 +42,107 @@
...
+      <?php
+      if ($registry['sql_run'])
+        echo "<span class='text'>" . xlt('DB installed') . "</span>";
+      else
+        echo "<a class='link_submit' href='./forms_admin.php?id={$registry['id']}&method=install_db'>".xl('install DB')."</a>";
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: attr() xlt()

File: interface/forms_admin/forms_admin.php [hunk 10]
```diff
@@ -37,103 +42,107 @@
...
+      ?>
+    </td>
+    <?php
+      echo "<td><input type='text' size='4'  name='priority_" . $registry['id'] . "' value='" . $priority_category['priority'] . "'></td>";
+      echo "<td><input type='text' size='10' name='category_" . $registry['id'] . "' value='" . $priority_category['category'] . "'></td>";
+      echo "<td><input type='text' size='10' name='nickname_" . $registry['id'] . "' value='" . $priority_category['nickname'] . "'></td>";
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: attr() attr()
(for above 3 lines)

File: interface/forms_admin/forms_admin.php [hunk 11]
```diff
@@ -37,103 +42,107 @@
...
+      echo "<td>";
+      echo "<select name='aco_spec_" . $registry['id'] . "'>";
```
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: attr()

File: interface/forms_admin/forms_admin.php (hunk 2 above)
Comment (sunsetsystems) [inline interface/forms_admin/forms_admin.php]: That's what I get for fixing white space. :)

File: library/acl.inc [hunk 12]
```diff
@@ -835,4 +835,48 @@
...
+      $s .= "</optgroup>";
+    }
+    return $s;
+  }
+
+  // Permissions check for a specified encounter form type.
+  // Note $return_value may be an array of return values.
+  //
+  function acl_check_form($formdir, $user='', $return_value='') {
+    require_once(dirname(__FILE__) . '/registry.inc');
```
Comment (bradymiller) [inline library/acl.inc]: Just noting that including scripts within functions can sometimes cause difficult to debug scope issues.

File: library/registry.inc [hunk 13]
```diff
@@ -57,6 +57,11 @@ function getRegistryEntry ( $id, $cols = "*" )
 	return sqlQuery($sql);
 }
 
+function getRegistryEntryByDirectory($directory, $cols = "*") {
+	$sql = "select $cols from registry where directory = '$directory'";
```
Comment (bradymiller) [inline library/registry.inc]: sql bind $directory

**Comment by bradymiller** [comment]
Very nice to see access controls getting some attention. Review complete; just some minor points, otherwise looks good to come into the codebase.

File: interface/forms_admin/forms_admin.php (hunk 2 above)
Comment (bradymiller) [inline interface/forms_admin/forms_admin.php]: yep, white space and the ambitious security flags :)

**Comment by sunsetsystems** [comment]
Fixes done, PR applied. Thanks Brady for the input.

**Pull request merged**